### PR TITLE
C#: Upgrade Roslyn dependencies to 3.9

### PR DESCRIPTION
**Pull request by tamasvajk**
[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/907/)

### CODE REVIEW

**Review by hvitved — APPROVED** [review]

File: csharp/ql/test/library-tests/controlflow/guards/Implications.expected [hunk 1]
```diff
@@ -209,7 +209,6 @@
 | Collections.cs:68:13:68:13 | access to local variable x | non-null | Collections.cs:63:17:63:55 | call to method ToList | non-null |
 | Collections.cs:68:13:68:13 | access to local variable x | null | Collections.cs:63:17:63:55 | call to method ToList | null |
 | Collections.cs:74:35:74:41 | ... == ... | true | Collections.cs:74:35:74:35 | access to parameter s | non-null |
-| Collections.cs:74:35:74:41 | ... == ... | true | Collections.cs:74:40:74:41 | "" | non-null |
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/controlflow/guards/Implications.expected:212]: this is unexpected, what happened to `""`.

File: csharp/ql/test/query-tests/Security Features/CWE-020/RuntimeChecksbypass.expected [hunk 2]
```diff
@@ -1,4 +1,3 @@
-| RuntimeChecksBypass.cs:20:13:20:22 | $"..." | This write to $@ may be circumventing a $@. | RuntimeChecksBypass.cs:7:19:7:19 | f | f | RuntimeChecksBypass.cs:11:9:14:9 | if (...) ... | check |
```
Comment (tamasvajk) [inline csharp/ql/test/query-tests/Security Features/CWE-020/RuntimeChecksbypass.expected:1]: this is unexpected

**Comment by tamasvajk** [comment]
I think this PR in Roslyn: `https://github.com/dotnet/roslyn/pull/49953` introduced the change. Note that only `#nullable warning enable`, and its equivalent compilation level setting turns on nullability analysis. So, I think we can't do it without reporting the compilation warnings.

File: csharp/ql/test/query-tests/Security Features/CWE-020/RuntimeChecksBypass.cs [hunk 3]
```diff
@@ -17,8 +17,10 @@ public Test1(string v)
     [OnDeserializing]
     public void Deserialize()
     {
-        f = $"invalid";  // BAD
+        f = GetString();  // BAD, non-constant and non-object creation expr
```
Comment (tamasvajk) [inline csharp/ql/test/query-tests/Security Features/CWE-020/RuntimeChecksBypass.cs:20]: Interpolated strings can have constant values now. See `https://github.com/dotnet/roslyn/pull/45396`.

**Comment by tamasvajk** [comment]
There's a recurring problem in the differences job: we can't determine the type of some conditional expressions:
```
NEW: cs/extraction-message
Error was generated by C# extractor: Failed to determine type at Semmle.Extraction.Context.ModelError(SyntaxNode node, String msg) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction/Context.cs:line 387 at Semmle.Extraction.CSharp.Entities.ExpressionNodeInfo.get_Type() in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/ExpressionNodeInfo.cs:line 73 at Semmle.Extraction.CSharp.Entities.Expression..ctor(IExpressionInfo info) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expression.cs:line 27 at Semmle.Extraction.CSharp.Entities.Expression`1..ctor(ExpressionNodeInfo info) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expression`1.cs:line 16 at Semmle.Extraction.CSharp.Entities.Expressions.Factory.Create(ExpressionNodeInfo info) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Factory.cs:line 261 at Semmle.Extraction.CSharp.Entities.Expressions.ImplicitCast..ctor(ExpressionNodeInfo info) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs:line 17 at Semmle.Extraction.CSharp.Entities.Expressions.ImplicitCast.Create(ExpressionNodeInfo info) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs:line 96 at Semmle.Extraction.CSharp.Entities.Expression.Create(Context cx, ExpressionSyntax node, IExpressionParentEntity parent, Int32 child) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expression.cs:line 91 at Semmle.Extraction.CSharp.Entities.Expressions.VariableDeclaration.<>c__DisplayClass9_0.<CreateDeclarator>b__0() in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/VariableDeclaration.cs:line 154 at Semmle.Extraction.Context.Try(SyntaxNode node, ISymbol symbol, Action a) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction/Context.cs:line 447
```

**Review by hvitved — APPROVED** [review]

**Pull request merged**